matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/space/ResolveSpaceInfoTask.kt

```diff
@@ -28,22 +28,40 @@ internal interface ResolveSpaceInfoTask : Task<ResolveSpaceInfoTask.Params, Spac
             val maxDepth: Int?,
             val from: String?,
             val suggestedOnly: Boolean?
-//            val autoJoinOnly: Boolean?
     )
 }
 
 internal class DefaultResolveSpaceInfoTask @Inject constructor(
         private val spaceApi: SpaceApi,
         private val globalErrorReceiver: GlobalErrorReceiver
 ) : ResolveSpaceInfoTask {
-    override suspend fun execute(params: ResolveSpaceInfoTask.Params): SpacesResponse {
-        return executeRequest(globalErrorReceiver) {
+
+    private lateinit var params: ResolveSpaceInfoTask.Params
```

Comment:
we can avoid the mutable state by chaining the `params` into the `getStableSpaceHierarchy/getUnstableSpaceHierarchy`

there's also an argument for avoiding `lateinit` in places that don't have a strictly documented lifecycle as it can cause unexpected side effects on order changes, eg if the `this.params = params` was set after `getSpaceHierarchy`

Reply:
I considered chaining params as an argument into these private functions but deliberately chose to make params as a lateinit instance variable. My reasoning is as follows:

1) The class is very cohesive (a.k.a its instance variables are used completely across the class), thus making its intended purpose clearer
2) It reduces the cognitive effort of reading each function. `getSpaceHierarchy()` is much easier to read than `getSpaceHierarchy(params)` where `params` is repeated across every function
3) I'm aware that a slight change in ordering would cause this to crash, but the contents and intent of the class is clear enough that a contributor shouldn't be led to making this mistake

It's a step I think is explained quite well in Uncle Bob's Clean Code, but if we still think it would be better to do otherwise then I'm happy to change it

Reply:
I would lean towards the introduction of mutability and state to the class as cognitively heavier than adding function parameters. As soon as we introduce mutability, threading and ordering become sources of bugs, things that we can safely avoid by keeping our logic stateless (ignoring our potentially stateful class dependencies :sweat_smile:)


```kotlin
override suspend fun execute(params: ResolveSpaceInfoTask.Params): SpacesResponse {
    return executeRequest(globalErrorReceiver) {
        try {
            spaceApi.getSpaceHierarchy(params)
        } catch (error: Throwable) {
            spaceApi.getSpaceHierarchyUnstable(params)
        }
    }
}

private suspend fun SpaceApi.getSpaceHierarchy(params: ResolveSpaceInfoTask.Params) = this.getSpaceHierarchy(
        spaceId = params.spaceId,
        suggestedOnly = params.suggestedOnly,
        limit = params.limit,
        maxDepth = params.maxDepth,
        from = params.from
)

private suspend fun SpaceApi.getSpaceHierarchyUnstable(params: ResolveSpaceInfoTask.Params) = this.getSpaceHierarchyUnstable(
        spaceId = params.spaceId,
        suggestedOnly = params.suggestedOnly,
        limit = params.limit,
        maxDepth = params.maxDepth,
        from = params.from
)
```

with that said, for this class I don't have too strong of an opinion as the scope is restricted and it's just 1 property, I would personally avoid `vars` and `lateinits` unless they serve a functional purpose as kotlin has other ways to reduce boilerplate

Reply:
Fair enough. I pushed the change!